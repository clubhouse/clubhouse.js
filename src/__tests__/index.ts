import Client from '../index';

class TestFactory {
  requests: object[];

  constructor(requests: object[]) {
    this.requests = requests;
  }
  makeRequest(url: string, method?: string, body?: any): Promise<any> {
    this.requests.push({ url, method, body });
    return Promise.resolve({
      ok: true,
      json: () => Promise.resolve(body),
    });
  }
}

const makeClient = (factory: TestFactory) =>
  new Client(
    {
      baseURL: 'http://localhost:4001',
      version: 'beta',
    },
    factory,
  );

describe('#Client', () => {
  it('return a new instance with the correct defaults', () => {
    const client = Client.create('token value');
    expect(client.version).toBe('beta');
  });

  describe('.listProjects', () => {
    it('returns a list of projects with a clubhouse account', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.listProjects();

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.getProject', () => {
    it('requests a single project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.getProject(1234);

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.createProject', () => {
    it('creates a new project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.createProject({ name: 'test' });

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.updateProject', () => {
    it('updates a existing project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.updateProject(12, { name: 'test' });

      expect(requests).toMatchSnapshot();
    });
  });

  describe('.deleteProject', () => {
    it('deletes a existing project', async () => {
      const requests = [];
      const client = makeClient(new TestFactory(requests));

      await client.deleteProject(1);

      expect(requests).toMatchSnapshot();
    });
  });
});
