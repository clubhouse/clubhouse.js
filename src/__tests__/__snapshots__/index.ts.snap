// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Client .createProject creates a new project 1`] = `
Array [
  Object {
    "body": Object {
      "name": "test",
    },
    "method": "POST",
    "url": "http://localhost:4001/api/beta/projects",
  },
]
`;

exports[`#Client .deleteProject deletes a existing project 1`] = `
Array [
  Object {
    "body": undefined,
    "method": "DELETE",
    "url": "http://localhost:4001/api/beta/projects/1",
  },
]
`;

exports[`#Client .getProject requests a single project 1`] = `
Array [
  Object {
    "body": undefined,
    "method": undefined,
    "url": "http://localhost:4001/api/beta/projects/1234",
  },
]
`;

exports[`#Client .listProjects returns a list of projects with a clubhouse account 1`] = `
Array [
  Object {
    "body": undefined,
    "method": undefined,
    "url": "http://localhost:4001/api/beta/projects",
  },
]
`;

exports[`#Client .updateProject updates a existing project 1`] = `
Array [
  Object {
    "body": Object {
      "name": "test",
    },
    "method": "PUT",
    "url": "http://localhost:4001/api/beta/projects/12",
  },
]
`;
